runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java

```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
+                payloadTranslator =
+                    KnownTransformPayloadTranslator.KNOWN_PAYLOAD_TRANSLATORS.get(
+                        appliedPTransform.getTransform().getClass());
+            try {
+              // Override the transform using the transform service.
+              res =
+                  updateTransformViaTransformService(
+                      urn, appliedPTransform, payloadTranslator, pipeline, res);
+            } catch (IOException e) {
+              throw new RuntimeException(e);
+            }
+          }
+        }
+      }
+    }
+
     // Validate that translation didn't produce an invalid pipeline.
     PipelineValidator.validate(res);
     return res;
   }
 
+  private static int findAvailablePort() throws IOException {
+    ServerSocket s = new ServerSocket(0);
+    try {
+      return s.getLocalPort();
+    } finally {
+      s.close();
+      try {
+        // Some systems don't free the port for future use immediately.
+        Thread.sleep(100);
+      } catch (InterruptedException exn) {
+        // ignore
+      }
+    }
+  }
+
+  // Override the given transform to the version available in a new transform service.
+  private static <
+          InputT extends PInput,
+          OutputT extends POutput,
+          TransformT extends org.apache.beam.sdk.transforms.PTransform<InputT, OutputT>>
+      RunnerApi.Pipeline updateTransformViaTransformService(
+          String urn,
+          AppliedPTransform<
+                  PInput,
+                  POutput,
+                  org.apache.beam.sdk.transforms.PTransform<? super PInput, POutput>>
+              appliedPTransform,
+          TransformPayloadTranslator<
+                  org.apache.beam.sdk.transforms.PTransform<? super PInput, POutput>>
+              originalPayloadTranslator,
+          Pipeline pipeline,
+          RunnerApi.Pipeline runnerAPIpipeline)
+          throws IOException {
+    ExternalTranslationOptions externalTranslationOptions =
+        pipeline.getOptions().as(ExternalTranslationOptions.class);
+
+    // Config row to re-construct the transform within the transform service.
+    Row configRow = originalPayloadTranslator.toConfigRow(appliedPTransform.getTransform());
+    ByteStringOutputStream outputStream = new ByteStringOutputStream();
+    try {
+      RowCoder.of(configRow.getSchema()).encode(configRow, outputStream);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    // Java expansion serivice able to identify and expand transforms that includes the construction
+    // config provided here.
+    ExternalTransforms.ExternalConfigurationPayload payload =
+        ExternalTransforms.ExternalConfigurationPayload.newBuilder()
+            .setSchema(SchemaTranslation.schemaToProto(configRow.getSchema(), true))
+            .setPayload(outputStream.toByteString())
+            .build();
+
+    String serviceAddress = null;
```

Comment:
Start and connect to the service once and pass it in as an option rather than doing so for every transform individually.